openshift-tools/pkg/discover-etcd-initial-cluster/initial-cluster.go

```diff
@@ -105,110 +113,136 @@ func (o *DiscoverEtcdInitialClusterOptions) Validate() error {
-	var memberDir string
-	if strings.HasSuffix(o.DataDir, "member") {
-		memberDir = o.DataDir
-		o.DataDir = filepath.Dir(o.DataDir)
-	} else {
-		memberDir = filepath.Join(o.DataDir, "member")
+	var dataDirExists bool
+	// check if dataDir structure exists
+	_, err := os.Stat(filepath.Join(o.DataDir, "member/snap"))
+    if err != nil && !os.IsNotExist(err) {
+    	return err
 	}
-
-	memberDirExists := false
-	_, err := os.Stat(memberDir)
-	switch {
-	case os.IsNotExist(err):
-		// do nothing. This just means we fall through to the polling logic
-
-	case err == nil:
-		fmt.Fprintf(os.Stderr, "memberDir %s is present on %s\n", memberDir, o.TargetName)
-		memberDirExists = true
-
-	case err != nil:
-		return err
+	if err == nil {
+		fmt.Fprintf(os.Stderr, "dataDir %s is present on %s\n", o.DataDir, o.TargetName)
+		dataDirExists = true
 	}
 
 	client, err := o.getClient()
-	if err != nil && memberDirExists {
-		// we weren't able to get client and need to return based previous memberDir so we can restart.  This is the off and on again flow.
-		fmt.Fprintf(os.Stderr, "Couldn't get client, but memberDir %s is present on %s, err=%s. Returning with no error.\n", memberDir, o.TargetName, err)
-		fmt.Printf(o.TargetName)
+
+	// Condition: unknown membership but dataDir exists
+	// possible reasons for this condition.
+	// 1.) single node etcd cluster
+	// 2.) transient networking problem
+	// 3.) on and off flow
+	// Result: empty initial config
+	if err != nil && dataDirExists {
+		fmt.Fprintf(os.Stderr, "Couldn't get client, but dataDir %s is present on %s, err=%s. Returning with no error.\n", o.DataDir, o.TargetName, err)
 		return nil
-	} else if err != nil {
+	}
+	if err != nil {
 		return err
 	}
 	defer client.Close()
 
-	var targetMember *etcdserverpb.Member
-	var allMembers []*etcdserverpb.Member
+	target := url.URL{
+		Scheme: o.TargetPeerURLScheme,
+		Host:   fmt.Sprintf("%s:%s", o.TargetPeerURLHost, o.TargetPeerURLPort),
+	}
+
 	for i := 0; i < 10; i++ {
 		fmt.Fprintf(os.Stderr, "#### attempt %d\n", i)
-		targetMember, allMembers, err = o.checkForTarget(client)
 
-		for _, member := range allMembers {
-			fmt.Fprintf(os.Stderr, "      member=%v\n", stringifyMember(member))
+		// check member list on each iteration for changes
+		cluster, err := client.MemberList(context.TODO())
+		if err != nil {
+			fmt.Fprintf(os.Stderr,"member list request failed: %v", err)
+			continue
+		}
+		logCurrentMembership(cluster.Members)
+		targetMember, found := checkTargetMember(target,cluster.Members)
+
+		// Condition: unstarted member found
+		// This member is part of the cluster but has not yet started. We know this because the name is populated at
+		// runtime which this member does not have.
+		// Result: provide initial cluster value so etcd can communicate with peers during startup.
+		if found && targetMember.Name == "" {
+			fmt.Print(getInitialCluster(o.TargetName, targetMember, cluster.Members))
+			return nil
 		}
```

Comment:
In this case, if data-dir exists, shouldn't we archive it?

Reply:
We were invoking `archiveDataDir(memberDir)` in this scenario in the previous code: 

https://github.com/openshift/etcd/blob/af729a8c3a354aa7eee3e27eb10658173fda613d/openshift-tools/pkg/discover-etcd-initial-cluster/initial-cluster.go#L182-L185

Reply:
> In this case, if data-dir exists, shouldn't we archive it?
If dataDir does exist with the above we should archive